utils/ami-cleanup/ami_cleanup.py

```diff
@@ -0,0 +1,217 @@
+    # Deregister AMI
+
+    try:
+        if not dry_run:
+            ec2_client.deregister_image(ImageId=ami_id)
+            logger.info(f"AMI {ami_id} deregistered")
+        else:
+            logger.info(f"AMI {ami_id} would be deregistered if this wasn't a dry run")
+    except Exception as e:
+        logger.debug(f"Error while deregistering AMI {ami_id}")
+        logger.debug(e)
+
+    # Get a list of snapshots associated with the ami
+    snap_filters = [template.replace("AMI_ID", ami_id) for template in snap_filter_templates]
+    snapshot_list = []
+
+    for snap_filter in snap_filters:
+        try:
+            snapshot_list_for_filter = ec2_client.describe_snapshots(
+                Filters=[{"Name": "description", "Values": [snap_filter]}]
+            )["Snapshots"]
+        except Exception as e:
+            logger.error(f"Failed to get list of Snapshots for AMI ID {ami_id} : {e}")
+            raise
+        if len(snapshot_list_for_filter) > 0:
+            snapshot_list.append(snapshot_list_for_filter)
+
+    if len(snapshot_list) > 0:
+        # Flatten list of lists generated above
+        flat_snapshot_list = []
+        for sublist in snapshot_list:
+            for item in sublist:
+                flat_snapshot_list.append(item)
+
+        for snapshot in flat_snapshot_list:
+            snapshot_id = snapshot["SnapshotId"]
+            try:
+                if not dry_run:
+                    ec2_client.delete_snapshot(SnapshotId=snapshot_id)
+                    logger.info(f"Snapshot {snapshot_id} deleted")
+                else:
+                    logger.info(
+                        f"Snapshot {snapshot_id} would be deleted if this wasn't a dry run"
+                    )
+            except Exception as e:
+                logger.debug(e)
+    else:
+        logger.error(f"No snapshots found for AMI {ami_id}")
+
+
+def main():
+    args = get_arguments()
+    if args.aws_profile_list is None:
+        args.aws_profile_list = []
+    main_client = get_ec2_client(args.aws_profile, args.aws_region)
+    candidate_list = get_deletion_list(main_client, args.ami_prefix, args.keep_min)
+    aws_profile_list_to_check = [args.aws_profile] + args.aws_profile_list
+
+    deletion_list = {}
+    for profile in aws_profile_list_to_check:
+        deletion_list[profile] = []
+        ec2_client = get_ec2_client(profile, args.aws_region)
+        asg_client = get_asg_client(profile, args.aws_region)
+        for ami in candidate_list:
+            if not check_ami_is_used(ec2_client, asg_client, ami["ImageId"]):
+                deletion_list[profile].append(ami["ImageId"])
+        logger.debug(f"Deletion list for profile {profile}:")
+        logger.debug(deletion_list[profile])
+
+    deletion_list_total = deletion_list[aws_profile_list_to_check[0]]
+    for profile in deletion_list:
+        deletion_list_total = list(set(deletion_list_total) & set(deletion_list[profile]))
```

Comment:
As it is, it performs set on `deletion_list_total` before joining, I think it should be done afterwards, otherwise you can end up with some duplicate items after the last pass through the loop
```suggestion
        deletion_list_total = set(list(deletion_list_total & set(deletion_list[profile])))
```